src/transformers/models/omdet_turbo/modeling_omdet_turbo.py

```diff
@@ -0,0 +1,1916 @@
+        """
+        a: B x K x H
+        b: B x H x K
+        """
+        a = OmDetTurboDeformableTransformerDecoder._norm(a, dim=2)
+        b = OmDetTurboDeformableTransformerDecoder._norm(b, dim=1)
+        # Calculating the Loss
+        logit_scale = logit_scale.exp()
+        logits_per_image = logit_scale * torch.bmm(a, b)
+        return logits_per_image
+
+    @staticmethod
+    def cls_score(cls_type, cls_feature, class_proj, logit_scale):
+        if cls_type == "cosine":
+            class_logits = OmDetTurboDeformableTransformerDecoder._b_cosine(
+                cls_feature, class_proj, logit_scale
+            )  # 4 100 256 4 256 20
+        elif cls_type == "dot":
+            class_logits = torch.bmm(cls_feature, class_proj)  # 4 100 20
+        else:
+            raise Exception("Unknown cls type {}".format(cls_type))
+        return class_logits
+
+    @staticmethod
+    def _inverse_sigmoid(x, eps=1e-5):
+        x = x.clamp(min=0, max=1)
+        x1 = x.clamp(min=eps)
+        x2 = (1 - x).clamp(min=eps)
+        return torch.log(x1 / x2)
+
+    def forward(
+        self,
+        embed,  # decoder embeddings
+        refer_bbox,  # anchor
+        feats,  # image features
+        shapes,  # feature shapes
+        label_feats,  # label features
+        task_feats,
+        bbox_head,
+        score_head,
+        pos_mlp,
+        attn_mask=None,
+        padding_mask=None,
+    ):
+        output = embed
+        dec_bboxes = []
+        dec_cls = []
+        last_refined_bbox = None
+        refer_bbox = refer_bbox.sigmoid()
+        for i, layer in enumerate(self.layers):
+            output, task_feats = layer(
+                output, task_feats, refer_bbox, feats, shapes, padding_mask, attn_mask, pos_mlp(refer_bbox)
+            )
+
+            # refine bboxes, (bs, num_queries+num_denoising, 4)
+            refined_bbox = torch.sigmoid(bbox_head[i](output) + self._inverse_sigmoid(refer_bbox))
+
+            clas_proj = score_head[i](label_feats).permute(1, 2, 0)
+
+            if self.training:
+                dec_cls.append(self.cls_score(self.cls_type, output, clas_proj, self.logit_scale))
+                if i == 0:
+                    dec_bboxes.append(refined_bbox)
+                else:
+                    dec_bboxes.append(torch.sigmoid(bbox_head[i](output) + self._inverse_sigmoid(last_refined_bbox)))
+            elif i == self.eval_idx:
+                dec_cls.append(self.cls_score(self.cls_type, output, clas_proj, self.logit_scale))
+                dec_bboxes.append(refined_bbox)
+                break
+
+            last_refined_bbox = refined_bbox
+            refer_bbox = refined_bbox.detach() if self.training else refined_bbox
+
+        return torch.stack(dec_bboxes), torch.stack(dec_cls)
+
+
+class OmDetTurboPreTrainedModel(PreTrainedModel):
+    config_class = OmDetTurboConfig
+    base_model_prefix = "model"
+    main_input_name = "pixel_values"
+
+    def _init_weights(self, module):
```

Comment:
Just a reminder to be accurate with heads initialization, that might be crucial for finetuning in case there will be mismatch sizes with the original checkpoint (if training is going to be implemented)